core/src/main/java/org/infinispan/reactive/publisher/impl/LocalClusterPublisherManagerImpl.java

```diff
@@ -143,21 +142,24 @@ IntSet handleNullSegments(IntSet segments) {
       SegmentPublisherSupplier<R> cachePublisher = localPublisherManager.keyPublisher(handleNullSegments(segments),
             keysToInclude, null, includeLoader, DeliveryGuarantee.AT_MOST_ONCE, transformer);
 
-      Flowable<K> keyFlowable = keyPublisherFromContext(invocationContext, segments, keyPartitioner, keysToInclude);
-
       return new SegmentPublisherSupplier<R>() {
          @Override
          public Publisher<Notification<R>> publisherWithSegments() {
-            Publisher<Notification<R>> cacheData = cachePublisher.publisherWithSegments();
-            return Flowable.concat(Flowable.fromPublisher(transformer.apply(keyFlowable))
-                                           .map(Notifications::value),
-                                   cacheData);
+            Flowable<SegmentPublisherSupplier.Notification<CacheEntry<K, V>>> contextFlowable =
+                  notificationPublisherFromContext(invocationContext, segments, keyPartitioner, keysToInclude);
+
+            return Flowable.concat(contextFlowable.concatMap(notification ->
+                        Flowable.fromPublisher(transformer.apply(Flowable.just(notification.value().getKey())))
```

Comment:
> IMO this is just too heavyweight.

In what regard?

1. We don't even hit this code if the context is null or empty (this will be the majority of cases)
2. This is only done for each value in the context, which normally is a tiny percentage in comparison to the actual cache data. 
3. It only happens when asking for segment information when iterating

I had debated about using groupBy to group the context entries by segment at least, so then you limit the number of transformer calls to # of segments, but that seems overly complicated for something that should be a lot smaller as per 2 above.

I actually have another idea I could invert determining segments and mapping the key. This would in turn only require a single mapping invocation, let me see if I can get that going.

> Speaking of which, it's weird that the concat happens directly in publisherWithSegments() in LocalClusterPublisherManagerImpl, but only later in SegmentAwarePublisherImpl.buildInitialCommand() for ClusterPublisherManagerImpl.

This is because in ClusterPublisherManagerImpl it has to be done only once (it is invoked multiple times), where as LocalClusterPublisherManagerImpl it is guaranteed to be done once. The latter is less complex code imo since it is more close together.

> It would be great if we could remove references to the invocation context in the publisher managers and set the excluded keys + process the non-null context keys directly in DistributedCacheStream, but then again that change is probably better suited for #9678

I am not a fan of that personally as then we have very similar logic in possibly 2 additional places outside of the actual iteration code that does this (2 because we will have DistributedCacheStream and our new API).

Reply:
> I actually have another idea I could invert determining segments and mapping the key. This would in turn only require a single mapping invocation, let me see if I can get that going.

I wasn't able to get this working, given the fact the the transformer could have asynchronously published values or buffering which would cause issue with tracking the key to values, similar to the issue with KeyPublisherState that you commented on as well.

Reply:
>     1. We don't even hit this code if the context is null or empty (this will be the majority of cases)

True

>     2. This is only done for each value in the context, which normally is a tiny percentage in comparison to the actual cache data.

True again.

>     3. It only happens when asking for segment information when iterating

This is the one where I'm less clear, does this happen only when the application uses a `SegmentCompletionListener`, or do we decide whether to include segment information based on some other stuff as well?

> I had debated about using groupBy to group the context entries by segment at least, so then you limit the number of transformer calls to # of segments, but that seems overly complicated for something that should be a lot smaller as per 2 above.

Yes, it does sound complicated. My idea was to make the transformer apply both to a publisher and to an individual element. (Initially I was thinking publisher without segment info and publisher with segment info, but then I learned about KeyCompleted notifications.)

> This is because in ClusterPublisherManagerImpl it has to be done only once (it is invoked multiple times), where as LocalClusterPublisherManagerImpl it is guaranteed to be done once. The latter is less complex code imo since it is more close together.

The way I see it, `publisherWithSegments()` in `ClusterPublisherManagerImpl` is also invoked only once, so the additional additional complexity is only needed because the concat is at a lower level compared to `LocalClusterPublisherManagerImpl`.

> > It would be great if we could remove references to the invocation context in the publisher managers and set the excluded keys + process the non-null context keys directly in DistributedCacheStream, but then again that change is probably better suited for #9678
> 
> I am not a fan of that personally as then we have very similar logic in possibly 2 additional places outside of the actual iteration code that does this (2 because we will have DistributedCacheStream and our new API).

I was assuming we wouldn't allow entry/key publishers to participate in transactions in AsyncCache or MutinyCache impls, because transactions are thread-local-based.

But even if we do allow them, and DistributedCacheStream is not a good place for them, I still think it's a good idea to keep the context entry handling in just one place.

>> I actually have another idea I could invert determining segments and mapping the key. This would in turn only require a single mapping invocation, let me see if I can get that going.
> I wasn't able to get this working, given the fact the the transformer could have asynchronously published values or buffering which would cause issue with tracking the key to values, similar to the issue with KeyPublisherState that you commented on as well.

Ack, we don't want to introduce restrictions that we could easily forget about later.